From 390eaf20124d3855b27f8b73047663cb25662555 Mon Sep 17 00:00:00 2001
From: John Clark <inindev@gmail.com>
Date: Fri, 2 May 2025 20:53:53 -0400
Subject: [PATCH 10/12] rockchip: rk3576: Add support for ArmSoM Sige5 with
 Rockchip RK3576

Add support for the ArmSoM Sige5 carrier board, powered by the 8nm
Rockchip RK3576 SoC (8nm, quad-core Cortex-A72 @ 2.2GHz, quad-core
Cortex-A53 @ 1.8GHz, Mali-G52 MC3 GPU, 6 TOPS NPU). The board supports
up to 16GB LPDDR4X, dual Gigabit Ethernet, WiFi 6, Bluetooth 5.0, and
4K HDMI/MIPI DSI outputs.

Signed-off-by: John Clark <inindev@gmail.com>
---
 arch/arm/dts/rk3576-armsom-sige5-u-boot.dtsi | 18 ++++++
 arch/arm/mach-rockchip/rk3576/Kconfig        |  8 +++
 board/armsom/sige5-rk3576/Kconfig            | 12 ++++
 board/armsom/sige5-rk3576/MAINTAINERS        |  7 +++
 configs/sige5-rk3576_defconfig               | 61 ++++++++++++++++++++
 doc/board/rockchip/rockchip.rst              |  1 +
 include/configs/sige5-rk3576.h               | 15 +++++
 7 files changed, 122 insertions(+)
 create mode 100644 arch/arm/dts/rk3576-armsom-sige5-u-boot.dtsi
 create mode 100644 board/armsom/sige5-rk3576/Kconfig
 create mode 100644 board/armsom/sige5-rk3576/MAINTAINERS
 create mode 100644 configs/sige5-rk3576_defconfig
 create mode 100644 include/configs/sige5-rk3576.h

diff --git a/arch/arm/dts/rk3576-armsom-sige5-u-boot.dtsi b/arch/arm/dts/rk3576-armsom-sige5-u-boot.dtsi
new file mode 100644
index 00000000000..30df20aef3e
--- /dev/null
+++ b/arch/arm/dts/rk3576-armsom-sige5-u-boot.dtsi
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright (c) 2025 John Clark <inindev@gmail.com>
+ *
+ */
+
+#include "rk3576-u-boot.dtsi"
+
+/ {
+	aliases {
+		mmc0 = &sdhci;
+		mmc1 = &sdmmc;
+	};
+};
+
+&sdhci {
+	cap-mmc-highspeed;
+};
diff --git a/arch/arm/mach-rockchip/rk3576/Kconfig b/arch/arm/mach-rockchip/rk3576/Kconfig
index e9e59c4d6b5..6ca50bea15f 100644
--- a/arch/arm/mach-rockchip/rk3576/Kconfig
+++ b/arch/arm/mach-rockchip/rk3576/Kconfig
@@ -1,5 +1,13 @@
 if ROCKCHIP_RK3576
 
+config TARGET_SIGE5_RK3576
+	bool "ArmSoM Sige5 with Rockchip RK3576"
+	help
+	  Enable support for the ArmSoM Sige5, powered by the 8nm Rockchip RK3576
+	  SoC with a quad-core Cortex-A72 @ 2.2GHz, quad-core Cortex-A53 @ 1.8GHz,
+	  Mali-G52 MC3 GPU, and 6 TOPS NPU. Supports up to 16GB LPDDR4X, dual
+	  Gigabit Ethernet, WiFi 6, Bluetooth 5.0, and 4K HDMI/MIPI DSI outputs.
+
 config TARGET_OMNI3576_RK3576
 	bool "Luckfox Omni3576 Carrier Board with Core3576 Module"
 	help
diff --git a/board/armsom/sige5-rk3576/Kconfig b/board/armsom/sige5-rk3576/Kconfig
new file mode 100644
index 00000000000..45ff164313b
--- /dev/null
+++ b/board/armsom/sige5-rk3576/Kconfig
@@ -0,0 +1,12 @@
+if TARGET_SIGE5_RK3576
+
+config SYS_BOARD
+	default "sige5-rk3576"
+
+config SYS_VENDOR
+	default "armsom"
+
+config SYS_CONFIG_NAME
+	default "sige5-rk3576"
+
+endif
diff --git a/board/armsom/sige5-rk3576/MAINTAINERS b/board/armsom/sige5-rk3576/MAINTAINERS
new file mode 100644
index 00000000000..4c872f9083e
--- /dev/null
+++ b/board/armsom/sige5-rk3576/MAINTAINERS
@@ -0,0 +1,7 @@
+SIGE5-RK3576
+M:	John Clark <inindev@gmail.com>
+R:	Jonas Karlman <jonas@kwiboo.se>
+F:	board/armsom/sige5-rk3576
+F:	include/configs/sige5-rk3576.h
+F:	configs/sige5-rk3576_defconfig
+F:	arch/arm/dts/rk3576-armsom-sige5*
diff --git a/configs/sige5-rk3576_defconfig b/configs/sige5-rk3576_defconfig
new file mode 100644
index 00000000000..156760dda76
--- /dev/null
+++ b/configs/sige5-rk3576_defconfig
@@ -0,0 +1,61 @@
+CONFIG_ARM=y
+CONFIG_SKIP_LOWLEVEL_INIT=y
+CONFIG_COUNTER_FREQUENCY=24000000
+CONFIG_ARCH_ROCKCHIP=y
+CONFIG_DEFAULT_DEVICE_TREE="rockchip/rk3576-armsom-sige5"
+CONFIG_ROCKCHIP_RK3576=y
+CONFIG_TARGET_SIGE5_RK3576=y
+CONFIG_SYS_LOAD_ADDR=0x40c00800
+CONFIG_DEBUG_UART_BASE=0x2AD40000
+CONFIG_DEBUG_UART_CLOCK=24000000
+CONFIG_DEBUG_UART=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_DEFAULT_FDT_FILE="rockchip/rk3576-armsom-sige5.dtb"
+# CONFIG_DISPLAY_CPUINFO is not set
+CONFIG_DISPLAY_BOARDINFO_LATE=y
+CONFIG_SPL_MAX_SIZE=0x40000
+# CONFIG_SPL_RAW_IMAGE_SUPPORT is not set
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_GPT=y
+CONFIG_CMD_I2C=y
+CONFIG_CMD_MMC=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_CMD_DNS=y
+CONFIG_CMD_WGET=y
+CONFIG_CMD_REGULATOR=y
+# CONFIG_SPL_DOS_PARTITION is not set
+CONFIG_OF_SPL_REMOVE_PROPS="clock-names interrupt-parent assigned-clocks assigned-clock-rates assigned-clock-parents"
+CONFIG_NET_LWIP=y
+CONFIG_ROCKCHIP_GPIO=y
+CONFIG_SYS_I2C_ROCKCHIP=y
+CONFIG_LED=y
+CONFIG_LED_GPIO=y
+CONFIG_SUPPORT_EMMC_RPMB=y
+CONFIG_MMC_DW=y
+CONFIG_MMC_DW_ROCKCHIP=y
+CONFIG_MMC_SDHCI=y
+CONFIG_MMC_SDHCI_SDMA=y
+CONFIG_MMC_SDHCI_ROCKCHIP=y
+CONFIG_PHY_REALTEK=y
+CONFIG_DWC_ETH_QOS=y
+CONFIG_DWC_ETH_QOS_ROCKCHIP=y
+CONFIG_PHY_ROCKCHIP_INNO_USB2=y
+CONFIG_PHY_ROCKCHIP_NANENG_COMBOPHY=y
+CONFIG_PHY_ROCKCHIP_USBDP=y
+CONFIG_DM_PMIC=y
+CONFIG_PMIC_RK8XX=y
+CONFIG_REGULATOR_RK8XX=y
+CONFIG_PWM_ROCKCHIP=y
+CONFIG_BAUDRATE=1500000
+CONFIG_DEBUG_UART_SHIFT=2
+CONFIG_SYS_NS16550_MEM32=y
+CONFIG_SYSRESET_PSCI=y
+CONFIG_USB=y
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_EHCI_GENERIC=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_OHCI_GENERIC=y
+CONFIG_USB_DWC3=y
+CONFIG_USB_DWC3_GENERIC=y
+CONFIG_ERRNO_STR=y
diff --git a/doc/board/rockchip/rockchip.rst b/doc/board/rockchip/rockchip.rst
index c20e374d650..c486ff259d7 100644
--- a/doc/board/rockchip/rockchip.rst
+++ b/doc/board/rockchip/rockchip.rst
@@ -134,6 +134,7 @@ List of mainline supported Rockchip boards:
      - Radxa ROCK 3B (rock-3b-rk3568)
 
 * rk3576
+     - ArmSoM Sige5 (sige5-rk3576)
      - Firefly ROC-RK3576-PC (roc-pc-rk3576)
      - Luckfox Omni3576 (omni3576-rk3576)
 
diff --git a/include/configs/sige5-rk3576.h b/include/configs/sige5-rk3576.h
new file mode 100644
index 00000000000..65820a984c0
--- /dev/null
+++ b/include/configs/sige5-rk3576.h
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright (c) 2024 ArmSoM Technology Co., Ltd.
+ */
+
+#ifndef __SIGE5_RK3576_H
+#define __SIGE5_RK3576_H
+
+#define ROCKCHIP_DEVICE_SETTINGS \
+		"stdout=serial,vidconsole\0" \
+		"stderr=serial,vidconsole\0"
+
+#include <configs/rk3576_common.h>
+
+#endif /* __SIGE5_RK3576_H */
-- 
2.39.5

